ARCs/arc-0035.md

```diff
@@ -0,0 +1,119 @@
+---
+arc: 35
+title: Algorand Offline Wallet Backup Protocol
+description: Wallet-agnostic backup protocol for multiple accounts
+author: Yigit Guler (@yigitguler)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/156
+status: Draft
+type: Standards Track
+category: Interface
+created: 2023-01-03
+---
+# Algorand Secure Offline Wallet Backup Protocol
+
+
+## Abstract
+This document outlines the high-level requirements for a wallet-agnostic backup protocol that can be used across all wallets on the Algorand ecosystem.
+
+## Specification
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+### Requirements
+At a high-level, offline wallet backup protocol has the following requirements:
+
+ * Wallet applications should allow backing up and storing multiple accounts at the same time.
+Account information should be encrypted with a user defined secret key, utilizing NaCl SecretBox method (audited and endorsed by Algorand).
+ 
+ * Encrypted final string should be easily copyable to be stored digitally.
+When importing, wallet applications should be able to detect already imported accounts and gracefully ignore them.
+
+### Format
+
+Before encryption, account information should be converted to the following JSON format:
+
+```
+{
+  "device_id": "UNIQUE IDENTIFIER FOR DEVICE (OPTIONAL)",
+  "provider_name": "PROVIDER NAME (OPTIONAL, i.e. Pera Wallet)",
+  "accounts": [
+    {
+      "address": "ACCOUNT PUBLIC ADDRESS (REQUIRED)"
+      "name": "USER DEFINED ACCOUNT NAME (OPTIONAL)",
+      "account_type": "TYPE OF ACCOUNT: single, multisig, watch, contact, ledger (REQUIRED)",
+    "private_key": "ACCOUNT PRIVATE KEY AS BASE64 ENCODING OF 64 BYTE PRIVATE KEY as encoded by NACL/libsodium (NOT PASSPHRASE, REQUIRED for user-owned accounts, can be omitted in case of watch, contact, ledger accounts)"
+      "metadata": "ANY ADDITIONAL CONTENT (OPTIONAL)",
```

Comment:
In the case of the ledger account, where should I include the path or the accountId? As an object in metadata or would a new field be included?

Reply:
Hi @yuli-ferna, sorry for the late reply.

Wouldn't address field work for this? I may be missing something, please let me know if I misunderstood the accountID.

Reply:
In order to sign a txn via Ledger Nano I need the account index (accountID) to indicate the device which account use to sign. For example:
Account Index #1 reference Account #1
Account Index #2 reference Account #2
Account Index #n reference Account #n

This would be what I mean when I say accountID value, it's a number that would be a separate value from the address.

Reply:
Yes, now I see. Thank you. @k13n how you are planing to approach this issue?

Reply:
hey @yigitguler, @yuli-ferna. We (Defly) store what you call the accountID in the `metadata` object, but refer to that as the `ledger_index`.

Perhaps it'd be a good idea to put together a list of common attributes that all wallets need to know and agree on certain names for them such that wallets can import and handle them appropriately.

Here's a list of attributes that we use in `metadata` and that might be useful for other wallets. Notice that for us these values are nullable but not optional (if an attribute doesn't apply, we keep the key in the object, but set its value to `null`).
- `ledger_connection_type` [int]: the type of the ledger (1=bluetooth/ble, 2=usb)
- `ledger_index` [int]: this is the index that hardware ledgers use to derive an account
- `ledger_device_id` [string]: this is the ID of the ledger
- `msig_signers` [string array]: the list of signers that constitute a multisig
- `msig_threshold`: [int]: the threshold of the multisig
- `msig_version` [int]: the version of multisig

Reply:
I think it would be good to document these in the ARC specification, otherwise, it can be impossible to parse some types of accounts or wallets may need to write.

@k13n is there a reason why ledger_connection_type is an int? It can be more error-proof if it is a string.

If you receive a backup file without these keys, does the app assumes that these are null or treat this as an error?

Reply:
@yigitguler, we just noticed that in the latest version of the ARC there's a `multisig` key in the account object that already contains what we called above the `msig_signers`, `msig_threshold`, and `msig_version`. That means we don't need this data in the `metadata` field. Should we add a similar object for ledger devices, something like this:

```json
{
  "derivation_index": "A LEDGER DEVICE CAN DERIVE MULTIPLE ALGORAND ACCOUNTS, THIS INDEX DENOTES THE N-TH SUCH ACCOUNT (REQUIRED, NUMBER FIELD)",
  "connection_type": "HOW THE LEDGER IS CONNECTED, EITHER bluetooth OR usb (OPTIONAL, STRING)",
  "device_id": "THE ID OF THE LEDGER DEVICE (OPTIONAL, STRING)"
}
```

Here's an example:

```json
{
  "device_id": "2498232091970170817",
  "provider_name": "Defly Wallet",
  "accounts": [
    {
      "address": "ELWRE6HZ7KIUT46EQ6PBISGD3ND6QSCBVWICYR2QR2Y7LOBRZRCAIKLWDE",
      "name": "My NFT Account",
      "account_type": "single",
      "ledger": {
        "device_id": "xyz",
        "index": 1,
        "connection_type": "bluetooth"
      }
    }
  ]
}
```

What do you think?

Reply:
Sounds great. I am updating the spec if you approve @OEA @miracozkan @yuli-ferna

Reply:
Waiting for @yuli-ferna's approval to close this issue.